### ISSUE TITLE
feat(generate): Add LogitProcessor trait for composable text generation

### ISSUE THREAD

**Issue by noahgift** [paiml/realizar#23]
## Summary

Add a `LogitProcessor` trait and `GenerationPipeline` to enable composable logit processing before token sampling. This is the missing abstraction that prevents downstream projects (like whisper.apr) from properly implementing token suppression, repetition penalties, and other generation constraints.

### Root Cause (from whisper.apr)

whisper.apr experiences hallucination (repetitive "the other one of the other one of...") because:

1. The decoder's `forward()` produces logits where SOT (start-of-transcript) has highest probability
2. There's no hook between `forward()` and `argmax` to suppress invalid tokens
3. The current `GenerationConfig` only supports sampling strategies, not logit manipulation

```
CURRENT (Broken):
  model.forward() → [logits] → argmax → token
                                  ↑
                          NO PROCESSING!
                  (SOT token wins, causes loop)

NEEDED (Fixed):
  model.forward() → [logits] → LogitProcessors → Sampler → token
                                     ↓
                          ├── TokenSuppressor
                          ├── RepetitionPenalty
                          └── TemperatureScaler
```

## Proposed API

```rust
// realizar/src/generate.rs

/// Context available during generation
pub struct GenerationContext<'a> {
    /// Previously generated tokens
    pub tokens: &'a [u32],
    /// Current generation step (0-indexed)
    pub step: usize,
    /// Vocabulary size
    pub n_vocab: usize,
}

/// Logit processor - composable pre-sampling transform
pub trait LogitProcessor: Send + Sync {
    /// Process logits in-place before sampling
    fn process(&self, logits: &mut [f32], ctx: &GenerationContext);
    
    /// Human-readable name for debugging
    fn name(&self) -> &str { "unnamed" }
}

/// Model trait for generation pipeline
pub trait GenerativeModel {
    /// Forward pass producing logits
    fn forward(&self, tokens: &[u32]) -> Result<Vec<f32>>;
    
    /// Vocabulary size
    fn vocab_size(&self) -> usize;
}

/// Generation pipeline with processor chain
pub struct GenerationPipeline<M: GenerativeModel> {
    model: M,
    processors: Vec<Box<dyn LogitProcessor>>,
    config: GenerationConfig,
}

impl<M: GenerativeModel> GenerationPipeline<M> {
    pub fn new(model: M) -> Self;
    pub fn add_processor(self, p: impl LogitProcessor + 'static) -> Self;
    pub fn with_config(self, config: GenerationConfig) -> Self;
    pub fn generate(&mut self, initial_tokens: &[u32]) -> Result<Vec<u32>>;
}
```

## Built-in Processors

```rust
/// Suppress specific tokens (set to -inf)
pub struct TokenSuppressor {
    suppress_ids: Vec<u32>,
}

/// Penalize repeated tokens/n-grams
pub struct RepetitionPenalty {
    penalty: f32,      // > 1.0 to penalize
    window: usize,     // Look-back window
}

/// Scale logits by temperature
pub struct TemperatureScaler {
    temperature: f32,  // > 1.0 = more random
}
```

## Use Case: whisper.apr

```rust
use realizar::generate::{GenerationPipeline, TokenSuppressor};

impl GenerativeModel for WhisperDecoder {
    fn forward(&self, tokens: &[u32]) -> Result<Vec<f32>> {
        self.forward_one(tokens, &self.encoder_output, &mut self.cache)
    }
    fn vocab_size(&self) -> usize { 51865 }
}

let pipeline = GenerationPipeline::new(decoder)
    .add_processor(TokenSuppressor::new(vec![SOT, PREV, SOLM]))  // Whisper-specific
    .with_config(GenerationConfig::greedy().with_eos_token_id(EOT));

let tokens = pipeline.generate(&initial_tokens)?;
```

## Acceptance Criteria

- [ ] `LogitProcessor` trait defined with `process()` and `name()` methods
- [ ] `GenerativeModel` trait defined with `forward()` and `vocab_size()` methods
- [ ] `GenerationContext` struct with token history and step
- [ ] `GenerationPipeline` struct with processor chain
- [ ] Built-in `TokenSuppressor` processor
- [ ] Built-in `RepetitionPenalty` processor
- [ ] Built-in `TemperatureScaler` processor
- [ ] Integration with existing `SamplingStrategy` enum
- [ ] Unit tests with >95% coverage
- [ ] Property-based tests for processor composition
- [ ] Documentation with examples

## Prior Art

- **HuggingFace Transformers**: `LogitsProcessor` and `LogitsProcessorList` (Python)
- **llama.cpp**: `llama_sample_*` functions for logit manipulation
- **vLLM**: `SamplingParams` with logit processors

## References

- Holtzman et al. (2020) "The Curious Case of Neural Text Degeneration" - Top-p sampling
- Wolf et al. (2020) "Transformers: State-of-the-Art NLP" - LogitsProcessor pattern
- whisper.apr spec: `docs/specifications/ground-truth-whisper-apr-cpp-hugging-face.md`
